### DISCUSSION TITLE
Get the public keys that signed a psbt from the script witness

### DISCUSSION THREAD

**Discussion by sebastianmontero** [rust-bitcoin/rust-bitcoin#1073] [Q&A] [answered]
Is there a way to get the public keys that signed a psbt from the script witness?

**Comment by sanket1729** [comment, marked as answer]
Unfortunately, this is complicated to solve in the general case. You would need to run the script interpreter and see which keys signed the script. 

The miniscript script interpreter has an API that returns the Satisfied constraints given script witness, and witness script. See: https://github.com/rust-bitcoin/rust-miniscript/blob/master/examples/verify_tx.rs for an example. 

Does that satisfy your use-case?

**Comment by sebastianmontero** [comment]
Thanks, so if understand correctly this is where the public keys that signed the transaction would be returned, right?: https://github.com/rust-bitcoin/rust-miniscript/blob/d589fe9338a87f476ff7d1bf84d91fbe75aea4d2/examples/verify_tx.rs#L62
What if the psbt does not have the witness script?

**Reply by sanket1729** [reply]
The witness script is the last element in the finalized witness.

**Reply by sanket1729** [reply]
In p2sh transactions, it is the last push opcode

**Comment by sebastianmontero** [comment]
Thanks
